libdevcrypto/sdf/SDFSM2Signature.cpp

```diff
@@ -0,0 +1,172 @@
+/*
+    This file is part of FISCO-BCOS.
+
+    FISCO-BCOS is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    FISCO-BCOS is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with FISCO-BCOS.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/** @file SDFSM2Signature.cpp
+ * @author maggiewu
+ * @date 2021-02-01
+ */
+#include "SDFSM2Signature.h"
+#include "SDFCryptoProvider.h"
+#include "csmsds.h"
+#include "libdevcore/Common.h"
+#include "libdevcore/FixedHash.h"
+#include "libdevcrypto/Common.h"
+#include "libdevcrypto/SM2Signature.h"
+#include "libdevcrypto/sm2/sm2.h"
+
+using namespace std;
+using namespace dev;
+using namespace dev::crypto;
+
+std::shared_ptr<crypto::Signature> dev::crypto::SDFSM2Sign(
+    KeyPair const& _keyPair, const h256& _hash)
+{
+    SDFCryptoProvider& provider = SDFCryptoProvider::GetInstance();
+    unsigned char signature[64];
+    unsigned int signLen;
+    Key key = Key();
+    key.setPrivateKey((unsigned char*)_keyPair.secret().ref().data(), 32);
+    key.setPublicKey((unsigned char*)_keyPair.pub().ref().data(), 64);
+    h256 privk((byte const*)key.PrivateKey(),
+        FixedHash<32>::ConstructFromPointerType::ConstructFromPointer);
+
+    // According to the SM2 standard
+    // step 1 : calculate M' = Za || M
+    // step 2 : e = H(M')
+    // step 3 : signature = Sign(e)
+    // get provider
+    BIGNUM* res = NULL;
```

Comment:
Please abstract this option to a single function, and hared with related calls of `SM2`